viewer-prototype/src/browser/trace-viewer/components/utils/filtrer-tree/tree.tsx

```diff
@@ -74,20 +98,30 @@ export class FilterTree extends React.Component<FilterTreeProps> {
 
     handleCheck = (id: number): void => {
         let checkedIds: number[] = [];
-        const checkedNode = this.getNode(id);
+        const checkedNode = this.getNode(this.props.nodes, id);
         if (checkedNode) {
-            if (checkedNode.children.length) {
-                const childrenIds = this.getAllChildrenIds(checkedNode, []);
-                if (this.isNodeChecked(id)) {
-                    checkedIds = checkedIds.concat(childrenIds);
-                } else {
-                    const childIdsToCheck = childrenIds.filter(childId => !this.isNodeChecked(childId));
+            const childrenIds = this.getAllChildrenIds(checkedNode, []);
+            const visibleChildrenIds = childrenIds.filter((childId: number) =>this.getNode(this.state.filteredNodes, childId) !== undefined);
+            if (!this.isNodeChecked(id)) {
+                if (checkedNode.children.length) {
+                    const childIdsToCheck = visibleChildrenIds.filter(childId => !this.isNodeChecked(childId));
                     checkedIds = checkedIds.concat(childIdsToCheck);
+                } else {
+                    checkedIds = checkedIds.concat(id);
+                }
+                if (this.areAllSiblingsChecked(checkedNode) && !this.isNodeChecked(checkedNode.parentId)) {
+                    const parentsToCheck = this.getParentIdsToCheck(checkedNode.parentId, [], true);
+                    checkedIds = checkedIds.concat(parentsToCheck);
                 }
             } else {
-                checkedIds.push(id);
-                if (this.areAllSiblingsChecked(checkedNode)) {
-                    checkedIds.push(checkedNode.parentId);
+                if (checkedNode.children.length) {
+                    checkedIds = checkedIds.concat(visibleChildrenIds);
+                } else {
+                    checkedIds = checkedIds.concat(id);
+                }
+                if (this.isNodeChecked(checkedNode.parentId)) {
+                    const parentsToCheck = this.getParentIdsToCheck(checkedNode.parentId, [], false);
+                    checkedIds = checkedIds.concat(parentsToCheck);
                 }
             }
             this.props.onChecked(checkedIds);
```

Comment:
As discussed offline, rename the onChecked method to onToggleCheck to make everything much clearer! ;-)